From 750673b4546d89625d9fea8a9de4b935dc1c672a Mon Sep 17 00:00:00 2001
From: Mauro Rossi <issor.oruam@gmail.com>
Date: Mon, 9 Dec 2019 07:56:32 +0100
Subject: [PATCH 066/100] intel_idle: Disable C6N and C6S on Bay Trail (v2)

(v2) Use flag CPUIDLE_FLAG_UNUSABLE to disable C6N and C6S due to commit
ba1e78a1dc ("cpuidle: Drop disabled field from struct cpuidle_state")

Signed-off-by: Mauro Rossi <issor.oruam@gmail.com>

It seems that Bay Trail SoCs sometimes have issues waking from C6,
a lot of users even report Bay Trail devices only being stable
when passing intel_idle.max_cstate=1 to the kernel.

This commits disables the C6 states while leaving the C7 states
available so that the cores can still reach deep sleep states.

There are several indicators that this is part of the solution for
all the users who need to pass intel_idle.max_cstate=1:

1) The "VLP52 EOI Transactions May Not be Sent if Software
   Enters Core C6 During an Interrupt Service Routine" errata.

2) Several users who need intel_idle.max_cstate=1 indicate in bko109051
   (which has over 800 comments!) that using a shell script which
   disables C6N and C6S through sysfs allows them to remove
   intel_idle.max_cstate=1 and still have a stable system which does
   use the C7 states for power-saving.

BugLink: https://bugzilla.kernel.org/show_bug.cgi?id=109051
Signed-off-by: Hans de Goede <hdegoede@redhat.com>
---
 drivers/idle/intel_idle.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/idle/intel_idle.c b/drivers/idle/intel_idle.c
index 3273360f30f7..85c630cd21f3 100644
--- a/drivers/idle/intel_idle.c
+++ b/drivers/idle/intel_idle.c
@@ -254,7 +254,7 @@ static struct cpuidle_state byt_cstates[] __initdata = {
 	{
 		.name = "C6N",
 		.desc = "MWAIT 0x58",
-		.flags = MWAIT2flg(0x58) | CPUIDLE_FLAG_TLB_FLUSHED,
+		.flags = MWAIT2flg(0x58) | CPUIDLE_FLAG_UNUSABLE,
 		.exit_latency = 300,
 		.target_residency = 275,
 		.enter = &intel_idle,
@@ -262,7 +262,7 @@ static struct cpuidle_state byt_cstates[] __initdata = {
 	{
 		.name = "C6S",
 		.desc = "MWAIT 0x52",
-		.flags = MWAIT2flg(0x52) | CPUIDLE_FLAG_TLB_FLUSHED,
+		.flags = MWAIT2flg(0x52) | CPUIDLE_FLAG_UNUSABLE,
 		.exit_latency = 500,
 		.target_residency = 560,
 		.enter = &intel_idle,
-- 
2.17.1

